devices/HumanDataCollector/HumanDataCollector.cpp

```diff
@@ -0,0 +1,663 @@
+            yError() << "The viewed IHumanDynamics interface from " << deviceName << " polydriver device might not be ready";
+            return false;
+        }
+
+        yInfo() << LogPrefix << deviceName << "attach() successful";
+        pImpl->isDynamicsEstimatorDevice = true;
+    }
+
+    return true;
+}
+
+bool HumanDataCollector::detach()
+{
+    while (isRunning()) {
+        stop();
+    }
+
+    if (pImpl->isStateProviderDevice) {
+
+        if (!pImpl->basePoseDataPort.isClosed()) {
+            pImpl->basePoseDataPort.close();
+        }
+
+        if (!pImpl->baseVelocityDataPort.isClosed()) {
+            pImpl->baseVelocityDataPort.close();
+        }
+
+        if (!pImpl->stateJointNamesDataPort.isClosed()) {
+            pImpl->stateJointNamesDataPort.close();
+        }
+
+        if (!pImpl->jointPositionsDataPort.isClosed()) {
+            pImpl->jointPositionsDataPort.close();
+        }
+
+        if (!pImpl->jointVelocitiesDataPort.isClosed()) {
+            pImpl->jointVelocitiesDataPort.close();
+        }
+
+        if (!pImpl->comPositionDataPort.isClosed()) {
+            pImpl->comPositionDataPort.close();
+        }
+
+        if (!pImpl->comVelocityDataPort.isClosed()) {
+            pImpl->comVelocityDataPort.close();
+        }
+
+        if (!pImpl->comProperAccelerationInBaseFrameDataPort.isClosed()) {
+            pImpl->comProperAccelerationInBaseFrameDataPort.close();
+        }
+
+        if (!pImpl->comProperAccelerationInWorldFrameDataPort.isClosed()) {
+            pImpl->comProperAccelerationInWorldFrameDataPort.close();
+        }
+
+        pImpl->iHumanState = nullptr;
+        pImpl->isStateProviderDevice = false;
+
+    }
+
+    if (pImpl->isWrenchProviderDevice) {
+
+        if (!pImpl->wrenchMeasurementSourceNamesDataPort.isClosed()) {
+            pImpl->wrenchMeasurementSourceNamesDataPort.close();
+        }
+
+        if (!pImpl->wrenchMeasurementValuesDataPort.isClosed()) {
+            pImpl->wrenchMeasurementValuesDataPort.close();
+        }
+
+        pImpl->iHumanWrenchMeasurements = nullptr;
+        pImpl->isWrenchProviderDevice = false;
+    }
+
+    if (pImpl->isDynamicsEstimatorDevice) {
+
+        if (!pImpl->wrenchEstimateSourceNamesDataPort.isClosed()) {
+            pImpl->wrenchEstimateSourceNamesDataPort.close();
+        }
+
+        if (!pImpl->wrenchEstimateValuesDataPort.isClosed()) {
+            pImpl->wrenchEstimateValuesDataPort.close();
+        }
+
+        if (!pImpl->dynamicsJointNamesDataPort.isClosed()) {
+            pImpl->dynamicsJointNamesDataPort.close();
+        }
+
+        if (!pImpl->jointTorquesDataPort.isClosed()) {
+            pImpl->jointTorquesDataPort.close();
+        }
+
+        pImpl->iHumanWrenchEstimates = nullptr;
+        pImpl->iHumanDynamics = nullptr;
+        pImpl->isDynamicsEstimatorDevice = false;
+    }
```

Comment:
Just checking, what would happen if we close a port that has never been opened? Maybe asking the nameserver if it is close and then issue the close commands is more demanding than just send the close commands whatsoever. It would also remove all these verbose `if`s.

Reply:
> Just checking, what would happen if we close a port that has never been opened?

I am not certain 

> Maybe asking the nameserver if it is close and then issue the close commands is more demanding than just send the close commands whatsoever.

As this is in the `detach()` which is called before termination https://github.com/robotology/human-dynamics-estimation/issues/109, it should not be a problem and moreover I feel it is cleaner a cleaner way to go with closing the ports which are quite a lot already with running all the devices of `HDE`. 

If this is not a blocking comment, I will proceed with resolving it. @diegoferigo let me know :wink:

Reply:
Not blocking, I was just trying to shorten the code :) Unfortunately, being the BufferedPorts a templated type, they cannot be inserted in a vector. And if you downcast to a `Contactable`, you cannot use `isClosed`.